File: grind75/Dec31/981.go
type pair struct {
    t int
    v string
}

type TimeMap struct {
    m map[string][]pair
}


func Constructor() TimeMap {
    return TimeMap{map[string][]pair{}}
}


func (this *TimeMap) Set(key string, value string, timestamp int)  {
    p := pair{t: timestamp, v: value}
    this.m[key] = append(this.m[key], p)
}


func (this *TimeMap) Get(key string, timestamp int) string {
    p, ok := this.m[key]
    if !ok {
        return ""
    }
    for i := len(p)-1; i >= 0; i-- {
        if p[i].t <= timestamp {
            s := p[i].v
            return s
        }
    }
    return ""
}


/**
 * Your TimeMap object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Set(key,value,timestamp);
 * param_2 := obj.Get(key,timestamp);
 */

File: grind75/Dec31/981.py
from collections import defaultdict
class TimeMap:

    def __init__(self):
        self.d = defaultdict(list)

    def set(self, key: str, value: str, timestamp: int) -> None:
        self.d[key].append((timestamp, value))

    def get(self, key: str, timestamp: int) -> str:
        if key not in self.d:
            return ""
        for i in range(len(self.d[key])-1, -1, -1):
            if self.d[key][i][0] <= timestamp:
                return self.d[key][i][1]
        
        return ""

